### PR TITLE
GG-18615

### PR DESCRIPTION
**Pull request by sanpwc**
Failover connections support for JDBC thin driver implemented.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java [hunk 1]
```diff
@@ -865,23 +875,31 @@ JdbcResultWithIo sendRequest(JdbcRequest req) throws SQLException {
...
+
+                while (retryAttemptsLeft > 0) {
+                    JdbcThinTcpIo cliIo = null;
+
+                    ensureConnected();
+
+                    try {
+                        cliIo = (stickyIo == null || !stickyIo.connected()) ? cliIo(calculateNodeIds(req)) : stickyIo;
 
                 if (stmt != null && stmt.requestTimeout() != NO_TIMEOUT) {
```
Comment (AMashenkov) [inline modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java]: Style. Let's make left padding wider.

File: modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java [hunk 2]
```diff
@@ -1698,6 +1731,49 @@ private void updateAffinityCache(JdbcQueryExecuteRequest qryReq, JdbcResponse re
...
+    /**
+     * Calculates query retries count for given {@param req}.
+     *
+     * @param stickyIo sticky connection, if any.
+     * @param req Jdbc request.
+     *
+     * @return retries count.
+     */
+    private int calculateRetryAttemptsCount(JdbcThinTcpIo stickyIo, JdbcRequest req) {
+        if (! affinityAwareness)
```
Comment (AMashenkov) [inline modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java]: Extra space in "!afinityAwareness"

File: modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java [hunk 3]
```diff
@@ -1698,6 +1731,49 @@ private void updateAffinityCache(JdbcQueryExecuteRequest qryReq, JdbcResponse re
...
+            req.type() == JdbcRequest.META_SCHEMAS ||
+            req.type() == JdbcRequest.CACHE_PARTITIONS)
+            return DFLT_RETRIES_CNT;
+
+        if (req.type() == JdbcRequest.QRY_EXEC) {
+            JdbcQueryExecuteRequest qryExecReq = (JdbcQueryExecuteRequest)req;
+
+            String trimmedQry = qryExecReq.sqlQuery().trim();
+
+            char[] qryCArr = trimmedQry.toCharArray();
```
Comment (AMashenkov) [inline modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java]: There is no need to have a copy char[]. String.indexOf() can be used instead.
BTW, it is possible to get false positive result if ';' is inside some string const.

File: modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java (hunk 1 above)
Comment (sanpwc) [reply modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java (hunk 2 above)
Comment (sanpwc) [reply modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java (hunk 3 above)
Comment (sanpwc) [reply modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java]: >There is no need to have a copy char[]. String.indexOf() can be used instead.
Yep, fixed.
>BTW, it is possible to get false positive result if ';' is inside some string const.
That's true. We either should have fully implemented parser on client side or use very simple rules with such false-positive effects. During the discussion, it was decided to start with the simpler version.

**Comment by dspavlov** [comment]
The ticket is resolved, closing PR.  
Feel free to reopen this PR if it is anyway actual.

**Pull request closed without merging**